src/main/java/io/causallabs/mustache/Feature.java

```diff
@@ -18,17 +18,45 @@ public interface Feature {
     public String getWarehouseTableName();
 
     /**
-     * return all the fields that are in the warehouse table for this feature
-     * 
-     * @return
+     * All the fields that are in the warehouse table for this feature
      */
     public Collection<WarehouseColumn> getWarehouseColumns();
 
     /**
-     * Return the events that are declared inside this feature
-     * 
-     * @return
+     * The events that are declared inside this feature
      */
     public Collection<FeatureEvent> getFeatureEvents();
 
+    /**
+     * The source code location where this feature was defined
+     */
+    public FileLocation getFileLocation();
+
+    /**
+     * The attributes of this feature marked with @per (i.e. are available to use as a per in a
+     * metric)
+     */
+    public Collection<WarehouseColumn> getPerColumns();
+
+    /**
+     * The attributes of this feature that are marked @session_key. Only applicable to sessions
+     */
+    public Collection<WarehouseColumn> getSessionKeys();
+
+    /**
+     * The attributes of this feature that are marked @split_key. Only applicable to sessions.
+     * Currently limited to one key
+     */
+    public Collection<WarehouseColumn> getSplitKeys();
+
+    /**
+     * The attribute of this feature that is used as the persistent key. Only applicable to sessions
+     */
+    public WarehouseColumn getPersistentKey();
+
+    /**
+     * The attributes of this feature that are defined with @elapsed
+     */
+    public Collection<WarehouseColumn> getElapsedColumns();
```

Comment:
Elapsed takes a parameter. So this is going to need to change when we extend it with more values.

Reply:
NVM, I see what you did below